### ISSUE TITLE
Autocomplete in Vivaldi doesn't work out of the box / does not support OpenSearch

### ISSUE THREAD

**Issue by alexs77** [searxng/searxng#2688]
**Version of SearXNG, commit number if you are using on master branch and stipulate if you forked SearXNG**

`2023.8.28+bad692e0f`

**How did you install SearXNG?**

searxng-docker

**What happened?**

No autocomplete suggestions are shown in the address bar when typing text.

**How To Reproduce**

1. Start Vivaldi. Using v6.1.3035.302 on macOS
2. Go to a searxng instance. Eg. mine at https://searx.foobar.vip
3. Right click on the search box and add as search engine
4. In Vivaldi search settings at `vivaldi://settings/search/` set SearXNG as the default search engine
5. Go to the address bar and start typing, eg. "warum"

**Expected behavior**

Just like with Bing or Chrome, I'd expect to get suggestions shown.

**Screenshots & Logs**

Chrome:

![Search suggestions from Chrome](https://github.com/searxng/searxng/assets/781361/357beb04-c1f6-4825-b78e-bcf2442d6b55)

**Additional context**

To get it to work, the auto complete ("URL für Suchvorschläge") needs to be filled correctly. To get the correct setting view the source code of the start page and search for `type="application/opensearchdescription+xml" rel="search"`. For me, it would be at `https://searx.foobar.vip/opensearch.xml?method=GET&autocomplete=google`.

In that file, it will be the line containing `rel="suggestions" type="application/x-suggestions+json"`; for me: 

```
    <Url rel="suggestions" type="application/x-suggestions+json" method="GET" template="https://searx.foobar.vip/autocompleter?q={searchTerms}"/>
```

And so the URL to enter will be: `https://searx.foobar.vip/autocompleter?q={searchTerms}`. This assumes, that in `settings.yml` autocomplete is set to something. Eg. `autocomplete: "google"`.

<img width="1189" alt="WORKING search settings" src="https://github.com/searxng/searxng/assets/781361/dd454dc5-42c7-48e3-96cd-bc0a417200ce">

**Comment by unixfox** [comment]
Related: https://github.com/searxng/searxng/issues/546

This used to work by default, not sure what changed since then in vivaldi.

**Comment by alexs77** [comment]
> Related: #546
> 
> This used to work by default, not sure what changed since then in vivaldi.

For completeness: When I right clicked and added it, the auto complete setting wasn't filled.

**Comment by return42** [comment]
Seems Vavaldi has some quirks / for reference I add some related posts:

- https://github.com/searxng/searxng/issues/546#issuecomment-987780347
- https://github.com/searxng/searxng/issues/431#issuecomment-975907274
- https://github.com/searxng/searxng/issues/2624

> Just like with Bing or Chrome, I'd expect to get suggestions shown.

Not sure if this is really an issue of SXNG or even Vivaldi

**Comment by alexs77** [comment]
> > Just like with Bing or Chrome, I'd expect to get suggestions shown.
> 
> Not sure if this is really an issue of SXNG or even Vivaldi

Well — I'd hope that SearXNG would work just as fine as Google or Bing and it should work just as fine in Chrome or Firefox.

The only thing which is constant here is SearXNG.

- Other search engines "work better" (as far as autocomplete in the browser is concerned).
- Other browsers "work better".

**Comment by return42** [comment]
With OpenSearch there should be no need to add a search engine manual / on matrix we described how to add a search engine in FFox .. 

To test something in your WEB browser or something in your SearXNG is wrong, try to add 

- https://de.wikipedia.org/wiki/Wikipedia:Hauptseite

as a search engine .. IDK how to do in Vivaldi but in FFox you find this lens with green + icon:

![image](https://github.com/searxng/searxng/assets/554536/2e725f52-66a4-4198-ba7c-9bae62161d0b)

Hint if you have added a search engine (in FFox) , like shown above the lens with the green plus will disappear (you can't add the same search engine twice).

**Comment by alexs77** [comment]
Yes, with Firefox or Chrome, there are no issues as far as SearXNG is
concerned.
Google or Bing also work fine in Vivaldi.

But not with Vivaldi and SearXNG.

**Comment by return42** [comment]
> Google or Bing also work fine in Vivaldi. But not with Vivaldi and SearXNG.

Just to confirm, in Vivaldi you can add 

- https://de.wikipedia.org/wiki/Wikipedia:Hauptseite

similar as described above for FFox, but you can't add 

- https://darmarit.org/searx/

that way in Vivaldi?

**Comment by alexs77** [comment]
Yes, it is possible to add the search engine. This is not what the bug is
about or what I described on Matrix.

It is so, that autocomplete ("Suchvorschläge") doesn't work in the browser.

"URL für Suchvorschläge" is empty.

![image](https://github.com/searxng/searxng/assets/781361/fe226a2d-b515-4a15-b45d-59e284519165)

And so Vivaldi doesn't try autocomplete.

Steps for reproducing.

1. Install Vivaldi
2. Start Vivaldi
3. Make sure to have a new, blank profile (so that other settings do not
interfere)
4. Go to a SearXNG site, eg. https://searx.foobar.vip/
5. Right click on the search box of SearXNG
6. Select "Als Suchmaschine hinzufügen" (add search engine)
7. Select "Set as Default"
8. Go to the address bar of the browser and enter "something"

- With Google or Bing, you'll now get search suggestions in Vivaldi.
- With Firefox or Chrome, you'll now get search suggestions from SearXNG

**Comment by return42** [comment]
I installed Vivaldi to give it a try (OT: what a horrible browser)

> 6. Select "Als Suchmaschine hinzufügen" (add search engine)

This is not [OpenSearch](https://github.com/searxng/searxng/issues/431) !!  --> Vavalid does not offer [Autodiscovery of search plugins](https://developer.mozilla.org/en-US/docs/Web/OpenSearch#autodiscovery_of_search_plugins) based on the [OpenSearch protocol](https://github.com/dewitt/opensearch)

When you add a search engine like you described, you simply add the URL of a HTML form to Vivald, by example, when you try to add 

- https://de.wikipedia.org/wiki/Wikipedia:Hauptseite

as I suggested above, you get entry like this:

![image](https://github.com/searxng/searxng/assets/554536/0043e23d-a9dc-4531-a0a3-eb2c0ebbc79a)

as you can see the "Suggest URL" is empty and needs to be filled out by hand.

> With Google or Bing, you'll now get search suggestions in Vivaldi.

Because you use the pre-configured setting from Vivaldi (you did not add this engines by yourself) .. try to remove bing and google from Vivialdi's list, the go to google (or bing) and try to add the search engines by yourself .. you will see this is not possible .. and if it is possible the "Suggest URL" leafs blank --> no auto-completion.

## TL;DR;

Vivaldi does not support OpenSearch!

----

From the advertisement for the Vivaldi browser ..

*"Every time you open the browser, you will discover a new tool or tweak that will make your life easier."*

.. and has earned a top ranking in my list of "browsers that sucks" .. just below MS-edge and pre-installed browser on smartphones :P

**Comment by alexs77** [comment]
> I installed Vivaldi to give it a try (OT: what a horrible browser)

Well, it's at least nicer than Firefox. Where's the tab management without extensions? Where's the sidebar management without extensions? etc.pp.

> > 6. Select "Als Suchmaschine hinzufügen" (add search engine)
> 
> This is not [OpenSearch](https://github.com/searxng/searxng/issues/431) !! --> Vavalid does not offer [Autodiscovery of search plugins](https://developer.mozilla.org/en-US/docs/Web/OpenSearch#autodiscovery_of_search_plugins) based on the [OpenSearch protocol](https://github.com/dewitt/opensearch)

Might be. At the end, I really don't care - for me, it's a technical detail, that I, as a user, don't want to know or worry about. I would like to get it to work.

I trust your word that it's not OpenSearch. I have never claimed anyting at all regarding OpenSearch.

> When you add a search engine like you described, you simply add the URL of a HTML form to Vivald, by example, when you try to add
> 
> * https://de.wikipedia.org/wiki/Wikipedia:Hauptseite
> 
> as I suggested above, you get entry like this:
> 
> ![image](https://user-images.githubusercontent.com/554536/263721500-0043e23d-a9dc-4531-a0a3-eb2c0ebbc79a.png)
> 
> as you can see the "Suggest URL" is empty and needs to be filled out by hand.

Exactly. That's what this bug is about. When you follow the steps I've shown when opening the bug, the "Suggest URL" is empty and users need to fill it manually. With the value as shown when I opened the ticket.

> > With Google or Bing, you'll now get search suggestions in Vivaldi.
> 
> Because you use the pre-configured setting from Vivaldi (you did not add this engines by yourself) .. try to remove bing and google from Vivialdi's list, the go to google (or bing) and try to add the search engines by yourself .. you will see this is not possible .. and if it is possible the "Suggest URL" leafs blank --> no auto-completion.

Correct.

* Google and Bing work fine - out of the box, at least.
* Chrome and Opera also work with SearXNG.

> 
> ## TL;DR;
> Vivaldi does not support OpenSearch!

Okay.

This means, that "autocomplete" (suggestions) don't work with SearXNG on Vivaldi.

Not a nice user experience.

> From the advertisement for the Vivaldi browser ..
> 
> _"Every time you open the browser, you will discover a new tool or tweak that will make your life easier."_
> 
> .. and has earned a top ranking in my list of "browsers that sucks" .. just below MS-edge and pre-installed browser on smartphones :P

No wonder that you're a Firefox fan — it's on my place #1 of the browsers that suck. Edge, OTHO, is rather nice nowadays. ;-)

**Comment by return42** [comment]
> Might be. At the end, I really don't care - for me, it's a technical detail, that I, as a user, don't want to know or worry about. I would like to get it to work.

This is what OpenSearch was made for!

> I trust your word that it's not OpenSearch. I have never claimed anyting at all regarding OpenSearch.

You are not alone / the developers from Vivaldi also never heard about .. its a widely use standard to register search engines in WEB browsers .. its used by wikipedia, startpage and millions of other WEB sites that provides a search service.

> Exactly. That's what this bug is about. When you follow the steps I've shown when opening the bug, the "Suggest URL" is empty and users need to fill it manually. With the value as shown when I opened the ticket.

Absolutely correct .. but its a drawback of Vivaldi, that you can't make use of the *"millions of other WEB sites that provides a search service."*

I can't see where we can help .. that is why I closed the ticket and removed the "bug" tag.

**Comment by return42** [comment]
BTW: thanks a lot for opening an issue at Vivaldi :+1:

- https://forum.vivaldi.net/topic/89969/support-opensearch-search-engines

**Comment by unixfox** [comment]
While Vivaldi does not support opensearch we can in the about page write a link to a documentation where we explain how to install SearXNG on Vivaldi. This makes me remind me that we should do the same for browsers that do not support opensearch yet, like Firefox on Android.

This way at least people that are not technical can still benefit from the autocompletion of SearXNG on Vivaldi.

@alexs77 Would you be able to write a new documentation page for vivaldi about that?

**Comment by alexs77** [comment]
FWIW: I added a feature request on the Vivaldi forums @ [Support OpenSearch search engines
](https://forum.vivaldi.net/topic/89969/support-opensearch-search-engines). Adding a vote to it might not hurt :)

And for the archive: On the Vivaldi forums, there's the post [Guide | Search Suggestion URL
](https://forum.vivaldi.net/topic/16991/guide-search-suggestion-url/14) with a long list of search engines and what to use for the _Search Suggestion URL_.

**Comment by alexs77** [comment]
> @alexs77 Would you be able to write a new documentation page for vivaldi about that?

I basically already did, didn't I? :)

But, yes, sure, I can copy and paste the things I wrote to the proper place and make it conform to standards.

If you could give me just one or two pointers to get me started, that would be useful. Like - where's the file to be modified. Or what's important to know.

I'll be happy to help out.

**Comment by return42** [comment]
>  This makes me remind me that we should do the same for browsers that do not support opensearch yet, like Firefox on Android.

What we need is a Browser Add-on where the user can choose one of the services listed on https://searx.space

**Comment by unixfox** [comment]
> > @alexs77 Would you be able to write a new documentation page for vivaldi about that?
> 
> I basically already did, didn't I? :)
> 
> But, yes, sure, I can copy and paste the things I wrote to the proper place and make it conform to standards.
> 
> If you could give me just one or two pointers to get me started, that would be useful. Like - where's the file to be modified. Or what's important to know.
> 
> I'll be happy to help out.

I'm not an expert on contributing to the documentation, @return42 is. (to be fair, I don't quite like how the documentation is organized, it's too complicated for beginners, but it's another completely other subject haha :smile: )

So there is a documentation on how to contribute to the documentation: https://docs.searxng.org/dev/contribution_guide.html#documentation. Basically the documentation files are in https://github.com/searxng/searxng/tree/master/docs

**Comment by return42** [comment]
>  it's too complicated for beginners

It is an efficient tool for the development, it significantly reduces the effort required for development & documentation (a large part of the documentation is generated from data & sources) --> it is not made for beginners who want to type a two-line text and forget what have been wrote.

- [reST primer](https://docs.searxng.org/dev/reST.html)
- [template](https://raw.githubusercontent.com/searxng/searxng/master/docs/user/configured_engines.rst) for [Configured Engines](https://docs.searxng.org/user/configured_engines.html)
- ..

> I basically already did, didn't I? :)

More or less .. sure .. Please summarize this again in an orderly fashion and write it up [here in a "Show and tell" discussion](https://github.com/searxng/searxng/discussions/new?category=show-and-tell) / Thanks :+1:
